jupyterhub/templates/scheduler/deployment.yaml

```diff
@@ -0,0 +1,59 @@
+{{ if .Values.scheduler.enabled -}}
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  labels:
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+    component: scheduler
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+  name: {{ .Chart.Name }}-scheduler
+spec:
+  replicas: {{ .Values.scheduler.replicas }}
+  template:
+    metadata:
+      annotations:
+        # This lets us autorestart when the configmap changes!
+        checksum/config-map: {{ include (print $.Template.BasePath "/scheduler/configmap.yaml") . | sha256sum }}
+      labels:
+        chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+        component: scheduler
+        heritage: {{ .Release.Service }}
+        release: {{ .Release.Name }}
+    spec:
+      volumes:
+      - name: config
+        configMap:
+          name: {{ .Chart.Name }}-scheduler-config
+      containers:
+      - command:
+        - /usr/local/bin/kube-scheduler
+        - --address=0.0.0.0
+        - --leader-elect=true
+        - --scheduler-name={{ .Chart.Name }}-{{ .Release.Name }}-scheduler
+        - --lock-object-namespace={{ .Release.Namespace }}
+        - --lock-object-name={{ .Chart.Name }}-{{ .Release.Name }}-scheduler-lock
```

Comment:
### Research note
Is the `--leader-elect` is supposed to be enabled "[...] when running replicated components for high availability." and it is defaulting to true.

Reading in [configure multiple schedulers - section 3](https://kubernetes.io/docs/tasks/administer-cluster/configure-multiple-schedulers/#3-run-the-second-scheduler-in-the-cluster) they write that if you want to setup leader election, you must update the following...

```
--leader-elect
--lock-object-namespace
--lock-object-name
```

And that one must also add the name of the scheduler to the ClusterRole under rules.apiGroups.resourceNames.

Reply:
While reading in the [kube-scheduler documentation](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/) I understood it as the `--lock-object-name` should be an existing Endpoint object by default, but can also be a ConfigMap if we also pass `--leader-elect-resource-lock=configmaps`.

Since we don't have an endpoint for our scheduler, since we don't have a Service for it currently and I'm unaware of the need for one, we should specify the configmap we use.

```yaml
        command:
        - /usr/local/bin/kube-scheduler
        - --address=0.0.0.0
        - --scheduler-name=jupyterhub-scheduler
        - --policy-configmap=scheduler-config
        - --policy-configmap-namespace={{ .Release.Namespace }}
        - --leader-elect=true
        - --leader-elect-resource-lock=configmaps
        - --lock-object-name=scheduler-config
        - --lock-object-namespace={{ .Release.Namespace }}
        - --v=4
```